### ISSUE TITLE
Make `Literal` behavior non-strict by default

### ISSUE THREAD

**Issue by KotlinIsland** [pydantic/pydantic#9991]
# TLDR

- enums should be converted, even when strict/`Literal`
- `Literal` should have an easy way to enable conversions/shouldn't disable conversions in the first place

### Initial Checks

- [X] I have searched Google & GitHub for similar requests and couldn't find anything
- [X] I have read and followed [the docs](https://docs.pydantic.dev) and still think this feature is missing

### Description

- inspiration: #9968

often i want to have a restricted value (1 option from an enum for example), but don't want to disable conversions. intuitively, i would use `Literal`, but pydantic adds extra rules to `Literal` that disable conversions:
```py
TypeAdapter(int).validate_python("1") # 1
TypeAdapter(Literal[1]).validate_python("1") # error: Input should be 1
```

so it appears that `Literal` is treated almost the same as `Annotated[_, Strict()]`

personally, i don't like that `Literal` isn't just a restricted type like it is in the typing space. ideally, i would want this behaviour removed and put into a separate `pydantic.Exactly`, used like `Annotated[Literal[1, 2, 3], Exactly()]`. although, once that is done, what would be differentiating it from `Strict`? so ideally i would want this behavior removed entirely in favour of using `Strict`

otherwise can a simple to use `NonLiteral`/`EnableConversions` be added:

```py
TypeAdapter(Annotated[Literal[1], NonLiteral()]).validate_python("1") # 1
```

or even more simply, can `Strict(False)` make this scenario work properly?

```py
TypeAdapter(Annotated[Literal[1], Strict(False)]).validate_python("1") # 1
```

in the meantime, it is impossible to load an `Enum`/`StrEnum` from json with strict/`Literal` and i must use something like:
```py
def convert_type(type_: object) -> object:
    def validator(value: object) -> object:
        try:
            return TypeAdapter(type_).validate_python(value)
        except ValueError:
            return value

    return BeforeValidator(validator)

class E(Enum):
    k = "v"

class M(BaseModel):
    e: Annotated[Literal[E.k], convert_type(E)]
    
M(e="v")
```

### Affected Components

- [ ] [Compatibility between releases](https://docs.pydantic.dev/changelog/)
- [X] [Data validation/parsing](https://docs.pydantic.dev/concepts/models/#basic-model-usage)
- [ ] [Data serialization](https://docs.pydantic.dev/concepts/serialization/) - `.model_dump()` and `.model_dump_json()`
- [ ] [JSON Schema](https://docs.pydantic.dev/concepts/json_schema/)
- [ ] [Dataclasses](https://docs.pydantic.dev/concepts/dataclasses/)
- [ ] [Model Config](https://docs.pydantic.dev/concepts/config/)
- [ ] [Field Types](https://docs.pydantic.dev/api/types/) - adding or changing a particular data type
- [ ] [Function validation decorator](https://docs.pydantic.dev/concepts/validation_decorator/)
- [ ] [Generic Models](https://docs.pydantic.dev/concepts/models/#generic-models)
- [ ] [Other Model behaviour](https://docs.pydantic.dev/concepts/models/) - `model_construct()`, pickling, private attributes, ORM mode
- [ ] [Plugins](https://docs.pydantic.dev/) and integration with other tools - mypy, FastAPI, python-devtools, Hypothesis, VS Code, PyCharm, etc.

**Comment by Viicos** [comment]
Similar: https://github.com/pydantic/pydantic-extra-types/issues/197

**Comment by sydney-runkle** [comment]
I find myself generally opposed to this change - this would be pretty breaking in nature, and I think it's fair + expected that `Literal` is effectively strict.

We can let @samuelcolvin and / or @davidhewitt weigh in before we make a final decision about whether to put this on the roadmap.

**Comment by KotlinIsland** [comment]
@sydney-runkle sure `Literal` can be strict, that's understandable and I could imagine someone having a use case for that (but why not just use `Strict`?). my main usability problem is that currently there is no way to restrict values without disabling automatic conversions, and that it is currently impossible to load `Literal`/'restricted' `Enum`s/`StrEnum`s etc from json without writing a custom validator

**Comment by sydney-runkle** [comment]
@KotlinIsland,

Perhaps the enum validator changes I'm planning for v2.10 can help, at least on the enum front. Will ask the team about the strict vs lax literals, but I don't think this would change util v3.

**Comment by sydney-runkle** [comment]
Chatted with the team - we're not planning to make this change. Stay tuned though for changes to the enum validator in rust that should better support the behavior you're looking for.

Closing as not planned, for now.

**Comment by KotlinIsland** [comment]
I'm working on a pydantic plugin, maybe I can change how `Literal` works in that

**Comment by Viicos** [comment]
Reopening for consideration for v3.

**Comment by mvanderlee** [comment]
Here is a workaround using a custom type:

```python
from typing import Literal, Annotated, TYPE_CHECKING
from pydantic import BaseModel, BeforeValidator

class IntLiteralType:
    def __class_getitem__(cls, values: int | tuple[int, ...]):
        if isinstance(values, int):
            values = (values,)

        return Annotated[
            Literal[*values], 
            BeforeValidator(lambda x: int(x)),
        ]

if TYPE_CHECKING:
    IntLiteral = Annotated[Literal, ...]
else:
    IntLiteral = IntLiteralType


class Foo(BaseModel):
    bar: IntLiteral[1, 2, 3] | None

print(Foo.model_json_schema())

Foo.model_validate({"bar": "1"})
```

```python
{'properties': {'bar': {'anyOf': [{'enum': [1, 2, 3], 'type': 'integer'}, {'type': 'null'}], 'title': 'Bar'}}, 'required': ['bar'], 'title': 'Foo', 'type': 'object'}
Foo(bar=1)
```
